### PR TITLE
Added appContext in Metrics

### PR DESCRIPTION
**Pull request by RidhamShah**
**Completed**

- Updated Metric model to add context as an array
- Updated Migration file
- Update CRUD operations
- Added a new API to get metrics only having the given context.
- Updated stepper to show metrics belonging to the context
- Updated test cases for the changes
- Updated format of pre-created metrics from .env file

**Remaining**

- create a way to add context in the JSON editor of metrics. ( For this part a new issue crated for UI design - [ISSUE](https://github.com/CarnegieLearningWeb/UpGrade/issues/1444) )

### CODE REVIEW

**Comment by RidhamShah** [comment]
Can we confirm about the new metrics structure of env file

```
PRE_DEFINED_METRICS = [
     {
          metrics: [
               {
                    metric: totalTimeSeconds, 
                    datatype: 'continuous'
               },
               {
                    metric: 'workspaceCompletionStatus',
                    ...
               },
               {
                    groupClass: 'addWorkspace',
                    allowedKeys: ['level1', 'level2'],
                    ...
               }
          ],
          contexts: [ context1, context2 ]
     },
     {
          metrics: [ ... ]
          contexts: [ ... ]
     }
]
```

File: backend/packages/Upgrade/src/init/seed/initMetrics.ts [hunk 1]
```diff
@@ -8,7 +8,9 @@ export function InitMetrics(logger: UpgradeLogger): Promise<any> {
   // Init default metrics in system
   if (env.initialization.metrics) {
     try {
-      return metricService.saveAllMetrics(JSON.parse(env.initialization.metrics), logger);
+      return JSON.parse(env.initialization.metrics).map((metricData) => {
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/src/init/seed/initMetrics.ts:11]: This is not returning a Promise but an array of Promise. You should use `Promise.all` before returning

File: backend/packages/Upgrade/.env.example [hunk 2]
```diff
@@ -87,4 +87,4 @@ CLIENT_API_SECRET = secret
...
-METRICS = [{"metric": "totalTimeSeconds","datatype": "continuous"}, { "groupClass": "masteryWorkspace", "allowedKeys": [ "calculating_area_various_figures", "Compare_functions_diff_reps_quadratic" ], "attributes": [{ "metric": "timeSeconds", "datatype": "continuous"}]}]
+PRE_DEFINED_METRICS = [{"metrics":[{"metric":"totalTimeSeconds","datatype":"continuous"},{"metric":"workspaceCompletionStatus","datatype":"categorical","allowedValues":["GRADUATED","PROMOTED"]}],"contexts":["test-context1","test-context2"]},{"metrics":[{"groupClass":"addWorkspace","allowedKeys":["level1","level2"],"attributes":[{"metric":"workspaceCompletionStatus","datatype":"categorical","allowedValues":["GRADUATED","PROMOTED"]},{"metric":"timeSpent","datatype":"continuous"}]}],"contexts":["test-context3"]}]
```
Comment (VivekFitkariwala) [inline backend/packages/Upgrade/.env.example:90]: The name change should be done on the server on deployment

File: frontend/projects/upgrade/src/app/features/dashboard/home/components/metrics/metrics.component.ts [hunk 3]
```diff
@@ -85,13 +85,16 @@ export class MonitoredMetricsComponent implements OnInit, OnChanges, OnDestroy {
   ) {}
 
   optionsSub() {
+    console.log('this.currentContext:', this.currentContext, this.experimentInfo?.context);
```
Comment (VivekFitkariwala) [inline frontend/projects/upgrade/src/app/features/dashboard/home/components/metrics/metrics.component.ts:88]: Remove this console statement

**Review by VivekFitkariwala — CHANGES_REQUESTED** [review]

File: backend/packages/Upgrade/src/init/seed/initMetrics.ts (hunk 1 above)
Comment (RidhamShah) [reply backend/packages/Upgrade/src/init/seed/initMetrics.ts]: code updated

**Comment by zackcl** [comment]
@RidhamShah I'm getting the `uncaughtException: column "context" of relation "metric" contains null values` error when trying this branch. Should I clear my postgres database to make the branch work?

**Comment by zackcl** [comment]
@RidhamShah I've updated the `METRICS` in the `.env` file as the one in the `.env.example` file, but I don't see the pre-created metrics in the UI. Is there something I need to do in order to add the metrics defined in the `.env` file?

```
METRICS=[{"metrics":[{"metric":"totalTimeSeconds","datatype":"continuous"},{"groupClass":"masteryWorkspace","allowedKeys":["calculating_area_various_figures","Compare_functions_diff_reps_quadratic"],"attributes":[{"metric":"timeSeconds","datatype":"continuous"}]}],"contexts":["context_identifier_1"]}]
```

Screenshot 2024-05-07 at 11 49 02 AM

**Comment by zackcl** [comment]
Nevermind, I saw that you renamed the `METRICS` to `PRE_DEFINED_METRICS` in your PR.

**Review by VivekFitkariwala — APPROVED** [review]

**Pull request merged**